## 七月在线机器学习第九期第二阶段考试
#### 考试说明:
- 起止时间：请同学在2018年7月11日至7月18日期间完成，最晚提交时间（7月18日24时之前）结束，<b>逾期不接受补考,该考试分数计入平时成绩</b>
- 考试方式：请同学将该试卷进行复制后，改名为自己姓名后，如State2Exam-WangWei.ipynb，<b>移动</b>至/0.Teacher/Exam/Stage2/目录下后，进行答题。
- 注意事项：为确保同学们真正了解自身对本周课程的掌握程度，<font color=red><b>请勿翻阅，移动，更改</b></font>其它同学试卷。如发现按0分处理
- 请同学在下方同学姓名处填写自己的姓名，批改人和最终得分不用填写

- 同学姓名:苏收
- 批改人：   
- 最终得分:

<center><h1>####答卷开始####</h1></center>
<hr>

## 问答题(共5题，每题10分，共计50分)

#### 1. 请写出你了解的机器学习特征工程操作都有哪些，及它们的具体实现？

机器学习的特征工程操作，主要包括特征处理、和特征选择两个部分。   
其中，
#### 特征处理
根据特征类型（数值、类别、时间、文本、统计、组合型特征）有不同的特征处理方法，   
数值型特征：可以进行幅度缩放，计算统计参数、离散化成多个区间等处理方式；   
类别型特征：可以采用One-Hot向量编码，进行正交化；   
时间型特征：可以离散化，也可以抽取年、月、日、周等作为特征；   
文本型特征：可以采用词袋模型或n-gram方法切割文本，提取特征，还可以使用TF-idf特征；   
统计特征：  可以人为提取数据集的一些统计学规律，作为特征使用，如购物转化率、商品的热度、排名等统计信息；   
组合特征：  可以人为对多个特征进行组合，也可以采用树模型进行特征组合。

#### 特征选取
采用过滤方法、包裹型或嵌入型方法，选取最重要的部分特征。
过滤方法：评估特征与结果相关程度，并排序，选取排序靠前的特征。
包裹型：递归删除部分弱特征，直到准确率或auc大幅度下滑为止，或选取特征子集进行评估，选取最优子集
嵌入型：根据模型来分析特征的重要性（比如根据产生的模型权重等）。

#### 2.模型评估中的留一法，留出法，交叉验证分别是什么操作？

1、留出法（hold out）：将数据集 D 划分为训练集 S 和测试集 T；   
2、留一法：留一法是留出法的特例，即每次测试集 T 只留一个数据，剩下的作为训练集 S；  
3、交叉验证：将数据集D划分成训练集、验证集和测试集。  
   其中验证集用于参数和模型选择，可反复使用。
   测试集仅用于模型效果评估，只可使用一次。

#### 3.网格搜索交叉验证的作用是什么，并简述网格搜索交叉验证是如何运行的？

网格搜索交叉交叉验证可以系统的遍历多种参数组合，并通过交叉验证评估筛选出最优的参数组合。   
运行方式分成两步：  
1、网格搜索，人为给出需要修改的模型参数后，程序遍历各种参数组合，  
2、没产生一种参数组合，意味着生成了一个新的模型，对每个模型，将数据集分成训练集、验证集和测试集后评估模型，以  
选取最优模型。


#### 4.工业界上模型融合有三大类方式？试简述每类方式其思想？

##### 1、bagging ，随机森林
通过抽取数据子集，基于同一个算法，训练多个模型，最后通过投票或平均的方式产生预测结果。或基于不同算法，采用过个模型结果的投票或平均值。
##### 2、 stacking
使用多种模型（LR，SVM，DT等）预测的结果作为特征，训练一个新的模型（模型2），采用模型2的预测结果。
##### 3、 Adaboost 和GBDT
通过重复迭代和训练、给分错的样本分配更高的权重、多个简单模型融合的方式改善模型效果。

#### 5.Early Stopping指的是什么？与它相关的概念有哪些？它们在实际项目中如何被运用？

Early stopping便是一种迭代次数截断的方法来防止过拟合的方法，即在模型对训练数据集迭代收敛之前停止迭代来防止过拟合。   
与之相关的概念有损失函数，L1正则和L2正则等。  
##### 运用方式
early stopping：一种基本的做法是，在每一个Epoch结束时（一个Epoch集为对所有的训练数据的一轮遍历）计算validation data的accuracy，当accuracy不再提高时，就停止训练。  
L1和L2正则：是通过在损失函数中加入相应的惩罚项（正则项），防止模型变得复杂或模型参数幅度过大。

## 代码题(共1题，共计50分)

#### 1.使用XGBoost的sklearn接口，对KaggleCredit2数据完成信用卡欺诈项目的建模及分析
- 要求以不同参数设置xgboost运行并进行效果比对,记录，最后给出实验报告

- KaggleCredit2数据文件 位于/home/ml9/0.Teacher/Data，请勿复制或移动

###### 1.读取数据文件

In [1]:
import pandas as pd
pd.set_option('display.max_columns', 500)
import zipfile
with zipfile.ZipFile('KaggleCredit2.csv.zip', 'r') as z:
    f = z.open('KaggleCredit2.csv')
    data = pd.read_csv(f, index_col=0)
data.head()

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines   age  \
0                 1                              0.766127  45.0   
1                 0                              0.957151  40.0   
2                 0                              0.658180  38.0   
3                 0                              0.233810  30.0   
4                 0                              0.907239  49.0   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                   2.0   0.802982         9120.0   
1                                   0.0   0.121876         2600.0   
2                                   1.0   0.085113         3042.0   
3                                   0.0   0.036050         3300.0   
4                                   1.0   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                             13.0                      0.0   
1                              4.0                      0.0   
2                              2.0                      1.0   
3                              5.0                      0.0   
4                              7.0                      0.0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                           6.0                                   0.0   
1                           0.0                                   0.0   
2                           0.0                                   0.0   
3                           0.0                                   0.0   
4                           1.0                                   0.0   

   NumberOfDependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

##### 2.处理空数值|

In [2]:
data.isnull().sum(axis=0)

SeriousDlqin2yrs                           0
RevolvingUtilizationOfUnsecuredLines       0
age                                     4267
NumberOfTime30-59DaysPastDueNotWorse       0
DebtRatio                                  0
MonthlyIncome                              0
NumberOfOpenCreditLinesAndLoans            0
NumberOfTimes90DaysLate                    0
NumberRealEstateLoansOrLines               0
NumberOfTime60-89DaysPastDueNotWorse       0
NumberOfDependents                      4267
dtype: int64

In [3]:
data.dropna(inplace=True)
data.shape

(108648, 11)

##### 3.设置特征和标签，并划分训练集和测试集

In [4]:
y = data['SeriousDlqin2yrs']
X = data.drop('SeriousDlqin2yrs', axis=1)

In [5]:
import numpy as np
msk = np.random.rand(len(data))  < 0.8
X_train = X[msk]
X_test = X[~msk]

y_train = y[msk]
y_test = y[~msk]

In [6]:
print(X_train.shape,X_test.shape)
print(y_train.shape,y_test.shape)

(87022, 10) (21626, 10)
(87022,) (21626,)


##### 4.导入XGBClassfier，并采用默认参数进行训练

In [7]:
from xgboost.sklearn import XGBClassifier
from sklearn import metrics

In [8]:
xgbc = XGBClassifier(learning_rate=0.1,max_depth=5,subsample=0.8,
                   n_estimators=1000 )
xgbc.fit(X_train,y_train,eval_metric='auc')

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=1, gamma=0, learning_rate=0.1, max_delta_step=0,
       max_depth=5, min_child_weight=1, missing=None, n_estimators=1000,
       n_jobs=1, nthread=None, objective='binary:logistic', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
       silent=True, subsample=0.8)

In [9]:
print('训练集准确度：%f'%xgbc.score(X_train,y_train))

训练集准确度：0.957562


In [10]:
y_predict = xgbc.predict(X_test)
error_num = (y_predict != y_test).sum()
probability = 1-float(error_num) / len(y_test)
probability

0.9354943124017386

输出默认参数下的模型评估结果

In [11]:
from sklearn.metrics import  classification_report

In [12]:
print(classification_report(y_test,y_predict,target_names=None, digits=2))

             precision    recall  f1-score   support

          0       0.94      0.99      0.97     20189
          1       0.54      0.18      0.27      1437

avg / total       0.92      0.94      0.92     21626



##### 5.采用网格搜索，寻求xgboost最优的参数配置

In [13]:
from sklearn import cross_validation
from sklearn.grid_search import GridSearchCV

C:\Users\shousunny\AppData\Local\conda\conda\envs\tensorflow\lib\site-packages\sklearn\cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)
C:\Users\shousunny\AppData\Local\conda\conda\envs\tensorflow\lib\site-packages\sklearn\grid_search.py:42: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. This module will be removed in 0.20.
  DeprecationWarning)


##### 首先搜索了学习速度，和booster两个参数
获得最优参数learning_rate = 0.07，  
booster参数取gbliner是会报错，这里就只搜索学习速率了

In [14]:
param_test0 = {'learning_rate':[0.01*x for x in range(1,13,3)]}

In [15]:
gsearch0 = GridSearchCV(estimator=XGBClassifier( n_estimators=140, max_depth=3,
min_child_weight=1, gamma=0, subsample=0.8,             colsample_bytree=0.8,
 objective= 'binary:logistic', nthread=4,     scale_pos_weight=1, seed=27),
                       param_grid = param_test0,     scoring='roc_auc',n_jobs=4,iid=False, cv=5)

In [16]:
gsearch0.fit(X_train,y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=0.8, gamma=0, learning_rate=0.1, max_delta_step=0,
       max_depth=3, min_child_weight=1, missing=None, n_estimators=140,
       n_jobs=1, nthread=4, objective='binary:logistic', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=27, silent=True,
       subsample=0.8),
       fit_params={}, iid=False, n_jobs=4,
       param_grid={'learning_rate': [0.01, 0.04, 0.07, 0.1]},
       pre_dispatch='2*n_jobs', refit=True, scoring='roc_auc', verbose=0)

In [17]:
gsearch0.grid_scores_,gsearch0.best_params_,gsearch0.best_score_

([mean: 0.84845, std: 0.00460, params: {'learning_rate': 0.01},
  mean: 0.85458, std: 0.00538, params: {'learning_rate': 0.04},
  mean: 0.85675, std: 0.00559, params: {'learning_rate': 0.07},
  mean: 0.85709, std: 0.00564, params: {'learning_rate': 0.1}],
 {'learning_rate': 0.1},
 0.8570887204453717)

##### 其次搜索树的深度和min_child_weight两个参数

In [18]:
#测试，选取树的深度和min_child_weight参数，进行网格搜索
param_test1 = {'max_depth':[x for x in range(3,10,2)],
              'min_child_weight':[x for x in range(1,6,2)]}

In [19]:
gsearch1 = GridSearchCV(estimator=XGBClassifier(learning_rate =0.07, n_estimators=140, max_depth=5,
min_child_weight=1, gamma=0, subsample=0.8,             colsample_bytree=0.8,
 objective= 'binary:logistic', nthread=4,     scale_pos_weight=1, seed=27),
                       param_grid = param_test1,     scoring='roc_auc',n_jobs=4,iid=False, cv=5)

In [20]:
gsearch1.fit(X_train,y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=0.8, gamma=0, learning_rate=0.07, max_delta_step=0,
       max_depth=5, min_child_weight=1, missing=None, n_estimators=140,
       n_jobs=1, nthread=4, objective='binary:logistic', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=27, silent=True,
       subsample=0.8),
       fit_params={}, iid=False, n_jobs=4,
       param_grid={'min_child_weight': [1, 3, 5], 'max_depth': [3, 5, 7, 9]},
       pre_dispatch='2*n_jobs', refit=True, scoring='roc_auc', verbose=0)

In [21]:
gsearch1.grid_scores_,gsearch1.best_params_,gsearch1.best_score_

([mean: 0.85675, std: 0.00559, params: {'min_child_weight': 1, 'max_depth': 3},
  mean: 0.85687, std: 0.00563, params: {'min_child_weight': 3, 'max_depth': 3},
  mean: 0.85683, std: 0.00566, params: {'min_child_weight': 5, 'max_depth': 3},
  mean: 0.85675, std: 0.00586, params: {'min_child_weight': 1, 'max_depth': 5},
  mean: 0.85679, std: 0.00613, params: {'min_child_weight': 3, 'max_depth': 5},
  mean: 0.85696, std: 0.00613, params: {'min_child_weight': 5, 'max_depth': 5},
  mean: 0.85421, std: 0.00633, params: {'min_child_weight': 1, 'max_depth': 7},
  mean: 0.85443, std: 0.00620, params: {'min_child_weight': 3, 'max_depth': 7},
  mean: 0.85480, std: 0.00615, params: {'min_child_weight': 5, 'max_depth': 7},
  mean: 0.85031, std: 0.00629, params: {'min_child_weight': 1, 'max_depth': 9},
  mean: 0.85080, std: 0.00609, params: {'min_child_weight': 3, 'max_depth': 9},
  mean: 0.85136, std: 0.00594, params: {'min_child_weight': 5, 'max_depth': 9}],
 {'max_depth': 5, 'min_child_weight': 5

#####  进行了两轮搜索，发现n_estimators最佳参数为100，

In [22]:
param_test2 = {'n_estimators':[x for x in range(100,1600,300)]}

In [23]:
gsearch2 = GridSearchCV(estimator=XGBClassifier(learning_rate =0.07,  max_depth=5,
min_child_weight=5, gamma=0, subsample=0.8,             colsample_bytree=0.8,
 objective= 'binary:logistic', nthread=4,     scale_pos_weight=1, seed=27),
                       param_grid = param_test2,     scoring='roc_auc',n_jobs=4,iid=False, cv=5)

In [25]:
gsearch2.fit(X_train,y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=0.8, gamma=0, learning_rate=0.07, max_delta_step=0,
       max_depth=5, min_child_weight=5, missing=None, n_estimators=100,
       n_jobs=1, nthread=4, objective='binary:logistic', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=27, silent=True,
       subsample=0.8),
       fit_params={}, iid=False, n_jobs=4,
       param_grid={'n_estimators': [100, 400, 700, 1000, 1300]},
       pre_dispatch='2*n_jobs', refit=True, scoring='roc_auc', verbose=0)

In [26]:
gsearch2.grid_scores_,gsearch2.best_params_,gsearch2.best_score_

([mean: 0.85694, std: 0.00604, params: {'n_estimators': 100},
  mean: 0.85333, std: 0.00665, params: {'n_estimators': 400},
  mean: 0.84938, std: 0.00729, params: {'n_estimators': 700},
  mean: 0.84503, std: 0.00740, params: {'n_estimators': 1000},
  mean: 0.84145, std: 0.00796, params: {'n_estimators': 1300}],
 {'n_estimators': 100},
 0.856938234458168)

In [27]:
param_test3 = {'n_estimators':[x for x in range(10,110,30)]}

In [33]:
gsearch3 = GridSearchCV(estimator=XGBClassifier(learning_rate =0.07,  max_depth=5,
min_child_weight=5, gamma=0, subsample=0.8,             colsample_bytree=0.8,
 objective= 'binary:logistic', nthread=4,     scale_pos_weight=1, seed=27),
                       param_grid = param_test3,     scoring='roc_auc',n_jobs=4,iid=False, cv=5)

In [34]:
gsearch3.fit(X_train,y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=0.8, gamma=0, learning_rate=0.07, max_delta_step=0,
       max_depth=5, min_child_weight=5, missing=None, n_estimators=100,
       n_jobs=1, nthread=4, objective='binary:logistic', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=27, silent=True,
       subsample=0.8),
       fit_params={}, iid=False, n_jobs=4,
       param_grid={'n_estimators': [10, 40, 70, 100]},
       pre_dispatch='2*n_jobs', refit=True, scoring='roc_auc', verbose=0)

In [45]:
gsearch3.grid_scores_,gsearch3.best_params_,gsearch3.best_score_

([mean: 0.85083, std: 0.00405, params: {'n_estimators': 10},
  mean: 0.85427, std: 0.00483, params: {'n_estimators': 40},
  mean: 0.85621, std: 0.00554, params: {'n_estimators': 70},
  mean: 0.85694, std: 0.00604, params: {'n_estimators': 100}],
 {'n_estimators': 100},
 0.856938234458168)

##### gama参数调优
最优参数为gama=0

In [42]:
param_test4={'gamma':[i/10.0 for i in range(0,5)]}

In [43]:
gsearch4 = GridSearchCV(estimator=XGBClassifier(learning_rate =0.07,  max_depth=5,
min_child_weight=5, n_estimators=100, subsample=0.8,             colsample_bytree=0.8,
 objective= 'binary:logistic', nthread=4,     scale_pos_weight=1, seed=27),
                       param_grid = param_test4,     scoring='roc_auc',n_jobs=4,iid=False, cv=5)

In [44]:
gsearch4.fit(X_train,y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=0.8, gamma=0, learning_rate=0.07, max_delta_step=0,
       max_depth=5, min_child_weight=5, missing=None, n_estimators=100,
       n_jobs=1, nthread=4, objective='binary:logistic', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=27, silent=True,
       subsample=0.8),
       fit_params={}, iid=False, n_jobs=4,
       param_grid={'gamma': [0.0, 0.1, 0.2, 0.3, 0.4]},
       pre_dispatch='2*n_jobs', refit=True, scoring='roc_auc', verbose=0)

In [46]:
gsearch5.grid_scores_,gsearch4.best_params_,gsearch4.best_score_

([mean: 0.85694, std: 0.00604, params: {'gamma': 0.0},
  mean: 0.85693, std: 0.00588, params: {'gamma': 0.1},
  mean: 0.85691, std: 0.00591, params: {'gamma': 0.2},
  mean: 0.85681, std: 0.00579, params: {'gamma': 0.3},
  mean: 0.85670, std: 0.00597, params: {'gamma': 0.4}],
 {'gamma': 0.0},
 0.856938234458168)

In [48]:
param_test5 ={'subsample':[i/100.0 for i in range(75,90,5)],
'colsample_bytree':[i/100.0 for i in range(75,90,5)]}

In [49]:
gsearch5 = GridSearchCV(estimator=XGBClassifier(learning_rate =0.07,  max_depth=5,
min_child_weight=5, n_estimators=100,gamma=0.0,
 objective= 'binary:logistic', nthread=4,     scale_pos_weight=1, seed=27),
                       param_grid = param_test5,     scoring='roc_auc',n_jobs=4,iid=False, cv=5)

In [50]:
gsearch5.fit(X_train,y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=1, gamma=0.0, learning_rate=0.07, max_delta_step=0,
       max_depth=5, min_child_weight=5, missing=None, n_estimators=100,
       n_jobs=1, nthread=4, objective='binary:logistic', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=27, silent=True,
       subsample=1),
       fit_params={}, iid=False, n_jobs=4,
       param_grid={'colsample_bytree': [0.75, 0.8, 0.85], 'subsample': [0.75, 0.8, 0.85]},
       pre_dispatch='2*n_jobs', refit=True, scoring='roc_auc', verbose=0)

In [51]:
gsearch5.grid_scores_,gsearch5.best_params_,gsearch5.best_score_

([mean: 0.85689, std: 0.00580, params: {'subsample': 0.75, 'colsample_bytree': 0.75},
  mean: 0.85649, std: 0.00558, params: {'subsample': 0.8, 'colsample_bytree': 0.75},
  mean: 0.85649, std: 0.00583, params: {'subsample': 0.85, 'colsample_bytree': 0.75},
  mean: 0.85674, std: 0.00609, params: {'subsample': 0.75, 'colsample_bytree': 0.8},
  mean: 0.85694, std: 0.00604, params: {'subsample': 0.8, 'colsample_bytree': 0.8},
  mean: 0.85643, std: 0.00613, params: {'subsample': 0.85, 'colsample_bytree': 0.8},
  mean: 0.85674, std: 0.00609, params: {'subsample': 0.75, 'colsample_bytree': 0.85},
  mean: 0.85694, std: 0.00604, params: {'subsample': 0.8, 'colsample_bytree': 0.85},
  mean: 0.85643, std: 0.00613, params: {'subsample': 0.85, 'colsample_bytree': 0.85}],
 {'colsample_bytree': 0.8, 'subsample': 0.8},
 0.856938234458168)

##### 正则参数搜索，
搜索出来都是默认参数，这有意思吗。。。

In [57]:
param_test6 = {
 'reg_alpha':[0, 0.001, 0.005, 0.01, 0.05]
}

In [58]:
gsearch6 = GridSearchCV(estimator=XGBClassifier(learning_rate =0.07,  max_depth=5,
min_child_weight=5, n_estimators=100,gamma=0.0,
 objective= 'binary:logistic', nthread=4,     scale_pos_weight=1, seed=27),
                       param_grid = param_test6,     scoring='roc_auc',n_jobs=4,iid=False, cv=5)

In [59]:
gsearch6.fit(X_train,y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=1, gamma=0.0, learning_rate=0.07, max_delta_step=0,
       max_depth=5, min_child_weight=5, missing=None, n_estimators=100,
       n_jobs=1, nthread=4, objective='binary:logistic', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=27, silent=True,
       subsample=1),
       fit_params={}, iid=False, n_jobs=4,
       param_grid={'reg_alpha': [0, 0.001, 0.005, 0.01, 0.05]},
       pre_dispatch='2*n_jobs', refit=True, scoring='roc_auc', verbose=0)

In [60]:
gsearch6.grid_scores_,gsearch6.best_params_,gsearch6.best_score_

([mean: 0.85620, std: 0.00596, params: {'reg_alpha': 0},
  mean: 0.85613, std: 0.00615, params: {'reg_alpha': 0.001},
  mean: 0.85607, std: 0.00625, params: {'reg_alpha': 0.005},
  mean: 0.85597, std: 0.00600, params: {'reg_alpha': 0.01},
  mean: 0.85614, std: 0.00600, params: {'reg_alpha': 0.05}],
 {'reg_alpha': 0},
 0.8562046501125616)

##### 6.在最优参数的基础上，降低学习速率，增加决策树的数量

In [61]:
xgbc1 = XGBClassifier(learning_rate=0.02,max_depth=5,subsample=0.8,
                   n_estimators=1000, min_child_weight=5)
xgbc1.fit(X_train,y_train,eval_metric='auc')

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=1, gamma=0, learning_rate=0.02, max_delta_step=0,
       max_depth=5, min_child_weight=5, missing=None, n_estimators=1000,
       n_jobs=1, nthread=None, objective='binary:logistic', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
       silent=True, subsample=0.8)

In [62]:
print('训练集准确度：%f'%xgbc1.score(X_train,y_train))

训练集准确度：0.942543


In [63]:
y_predict = xgbc1.predict(X_test)
error_num = (y_predict != y_test).sum()
probability = 1-float(error_num) / len(y_test)
probability

0.9371589753075003

In [64]:
print(classification_report(y_test,y_predict,target_names=None, digits=2))

             precision    recall  f1-score   support

          0       0.94      0.99      0.97     20189
          1       0.59      0.17      0.27      1437

avg / total       0.92      0.94      0.92     21626



##### 7.结果分析

与第四节中，采用默认参数的xgboost计算结果相比：   

Name | 训练集准确度 | 测试集准确度
- | :-: | -:
默认参数 | 0.95756 | 0.93549
调优参数 | 0.94254 | 0.93715 

从上表中可以看出，参数调优后，训练集准确度下降，测试集准确度上升了近2个百分点，可见模型泛化能力得到了一定提升。  

 采用默认参数的模型评估结果：
 
name|precision |   recall | f1-score |  support
- | :-: | -:
0    |   0.94  |    0.99  |    0.97  |   20189
1    |   0.54   |   0.18    |  0.27     | 1437
avg / total    |   0.92   |   0.94    |  0.92   |  21626
 
 参数调优后的模型评估结果：
 
 name|precision |   recall | f1-score |  support
- | :-: | -:
0    |   0.94  |    0.99  |    0.97  |   20189
1    |   0.59   |   0.17    |  0.27     | 1437
avg / total    |   0.92   |   0.94    |  0.92   |  21626

从模型评估结果来看，参数调优后对类别“1”，预测的准确度提升了，但召回率有所下降，其他评估结果没有显著变化。  
尝试了采用网格搜索方法，分批次的寻求最优参数，从网格搜索结果来看，大部分参数的最优解都是默认参数，  
可以认为收效并不显著。  
后续是不是可以通过特征工程、模型融合等其他方式提升模型性能？  

***
<center><h1>####答卷结束####</h1></center>